### PR TITLE
refactor(flightsql): Use upstream implementation of XdbcTypeInfo builder, support filter by datatype

### PR DESCRIPTION
**Pull request by alamb**
Similarly to https://github.com/influxdata/influxdb_iox/pull/8453

related to https://github.com/influxdata/influxdb_iox/issues/7569

This code has been ported upstream (see https://github.com/apache/arrow-rs/pull/4359 and https://github.com/apache/arrow-rs/pull/4340) so we can use the upstream implementation rather than our copy in IOx

It turns out that the upstream version supports filtering that our IOx version did not, meaning that is now supported too (kudos to @appletreeisyellow  for writing the test that shows this)

Test coverage is in https://github.com/influxdata/influxdb_iox/blob/3d972561d5467531939fc9463a0a90fce7d38cbe/influxdb_iox/tests/end_to_end_cases/flightsql.rs#L1

### CODE REVIEW

File: flightsql/src/planner.rs [hunk 1]
```diff
@@ -471,15 +470,10 @@ async fn plan_get_table_types(ctx: &IOxSessionContext) -> Result<LogicalPlan> {
 /// Return a `LogicalPlan` for GetXdbcTypeInfo
 async fn plan_get_xdbc_type_info(
     ctx: &IOxSessionContext,
-    data_type: Option<i32>,
+    cmd: CommandGetXdbcTypeInfo,
 ) -> Result<LogicalPlan> {
-    match data_type {
-        None => Ok(ctx.batch_to_logical_plan(TYPE_INFO_RECORD_BATCH.clone())?),
-        // TODO chunchun: support search by data_type
```
Comment (alamb) [inline flightsql/src/planner.rs:478]: Upstream code supports this searching 🎉

File: flightsql/src/xdbc_type_info/mod.rs [hunk 2]
```diff
@@ -56,200 +56,123 @@
...
-            None => create_params_builder.append(false), // create_params is nullable
-        }
+pub(crate) fn xdbc_type_info_data() -> &'static XdbcTypeInfoData {
+    &XDBC_TYPE_INFO_DATA
+}
+
+/// Data Types supported by DataFusion
+/// <https://arrow.apache.org/datafusion/user-guide/sql/data_types.html>
+static XDBC_TYPE_INFO_DATA: Lazy<XdbcTypeInfoData> = Lazy::new(|| {
+    let mut builder = XdbcTypeInfoDataBuilder::new();
```
Comment (alamb) [inline flightsql/src/xdbc_type_info/mod.rs:70]: This data is the same, but I had to update some of the types to reflect what is in the arrow-rs crate

File: influxdb_iox/tests/end_to_end_cases/flightsql.rs [hunk 3]
```diff
@@ -1174,15 +1174,24 @@ async fn flightsql_get_xdbc_type_info() {
...
+                    // test filter by type
                     let mut client = flightsql_client(state.cluster());
-                    // TODO chunchun: search by data_type test case
                     let data_type: Option<i32> = Some(6);
 
-                    let err = client.get_xdbc_type_info(data_type).await.unwrap_err();
-
-                    assert_matches!(err, FlightError::Tonic(..));
-                    assert_contains!(err.to_string(), "GetXdbcTypeInfo does not yet support filtering by data_type");
+                    let stream = client.get_xdbc_type_info(data_type).await.unwrap();
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/flightsql.rs:1181]: it works!

File: flightsql/src/xdbc_type_info/value.rs [hunk 4]
```diff
@@ -1,283 +0,0 @@
-use once_cell::sync::Lazy;
-
-/// [Arrow FlightSQL Specification]: https://github.com/apache/arrow/blob/9588da967c756b2923e213ccc067378ba6c90a86/format/FlightSql.proto#L948-L973
-///
-/// Note: Some of the data types are not supported by DataFusion yet:
-/// <https://arrow.apache.org/datafusion/user-guide/sql/data_types.html#unsupported-sql-types>
-#[derive(Debug, Clone)]
-#[allow(dead_code)]
-pub enum XdbcDataType {
```
Comment (alamb) [inline flightsql/src/xdbc_type_info/value.rs:9]: this is now all properly exposed upstream (thanks to https://github.com/apache/arrow-rs/pull/4340)

**Review by appletreeisyellow — APPROVED** [review]
Neat! Thank you for bringing the upstream XdbcTypeInfo builder! 💯

**Pull request merged**